libqtile/widget/bluetooth_extra.py

```
@@ -0,0 +1,634 @@
+
+    @catch_error("Unable to connect to device: {self.name}.")
+    async def connect(self):
+        await self.interface.call_connect()
+
+    @catch_error("Unable to disconnect device: {self.name}.")
+    async def disconnect(self):
+        await self.interface.call_disconnect()
+
+    @catch_error("Unable to pair device: {self.name}.")
+    async def pair_and_connect(self):
+        await self.interface.call_pair()
+        await self.connect()
+
+    async def action(self):
+        """Helper method to call appropriate method based on device status."""
+        if self.connected:
+            await self.disconnect()
+        elif self.paired and not self.connected:
+            await self.connect()
+        elif not self.paired:
+            await self.pair_and_connect()
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def connected(self):
+        return self._connected
+
+    @property
+    def paired(self):
+        return self._paired
+
+    @property
+    def status(self):
+        return self._status
+
+    @property
+    def battery(self):
+        if self.battery_device:
+            return self._battery
+        return ""
+
+    @property
+    def adapter(self):
+        a = self.widget.adapters.get(self._adapter)
+        if a:
+            return a.name
+        else:
+            return "Unknown"
+
+    def add_battery(self):
+        """Triggers adding battery interface."""
+
+        def refresh(_):
+            if self.battery_device:
+                create_task(self.update_props())
+
+        task = create_task(self.get_battery())
+        task.add_done_callback(refresh)
+
+    def remove_battery(self):
+        self.battery_device = None
+        self._battery = 0
+
+    async def get_battery(self):
+        proxy = await self.widget.get_proxy(self.path)
+        try:
+            self.battery_device = proxy.get_interface(BLUEZ_BATTERY)
+        except InterfaceNotFoundError:
+            return
```

There are a few ignored exceptions. Might be bit cleaner to use `contextlib.suppress`?